rust/src/transformations/count/make_count.tex

```diff
@@ -0,0 +1,152 @@
+    \end{itemize}
+\end{itemize}
+
+\subsection*{Implementation}
+\begin{lstlisting}[language = Python, escapechar=|]
+def make_count():
+    input_domain = VectorDomain(AllDomain(TIA))
+    output_domain = AllDomain(TO) |\label{line:output-domain}|
+
+    def function(data: Vec[TIA]) -> TO:|\label{line:TO-output}|
+        size = input_domain.size(data)
+        try: |\label{line:try-catch}|
+            return TO.exact_int_cast(size)
+        except FailedCast:
+            return TO.MAX_CONSECUTIVE |\label{line:except-return}|
+
+    input_metric = SymmetricDistance()
+    output_metric = AbsoluteDistance(TO)
+
+    stability_map = new_stability_map_from_constant(TO.one()) |\label{line:stability-map}|
+
+    return Transformation(
+        input_domain, output_domain, function,
+        input_metric, output_metric, stability_map)
+\end{lstlisting}
+
+\subsection*{Postcondition}
+\texttt{make\_count} raises an exception or returns a valid \texttt{Transformation}.
+
+\section{Proof}
+
+\transformationTheorem{\texttt{(TIA, TO)}}{\texttt{make\_count}}
+
+\begin{proof} \textbf{(Part 1 -- appropriate output domain).}
+    The \texttt{output\_domain} is \texttt{AllDomain(TO)}, so it is sufficient to show that \texttt{function} always returns non-null values of type \texttt{TO}.
+    By the definition of the \texttt{ExactIntCast} trait, \texttt{TO.exact\_int\_cast} always returns a non-null value of type \texttt{TO} or raises an exception.
+    If an exception is raised, the function returns \texttt{TO.MAXIMUM\_CONSECUTIVE}, which is also a non-null value of type \texttt{TO}.
+    Thus, in all cases, the function (from line \ref{line:try-catch}) returns a non-null value of type \texttt{TO}.
+\end{proof}
+
+\begin{proof} \textbf{(Part 2 -- domain-metric compatibility).}
+    Our \texttt{input\_metric} of \rustdoc{metrics}{struct}{SymmetricDistance} is compatible with any domain of the form \texttt{VectorDomain(inner\_domain)}, 
+    and our \texttt{input\_domain} of \\\texttt{VectorDomain(AllDomain(TIA))} is of this form. 
+    Therefore our \texttt{input\_domain} and \texttt{input\_metric} are compatible.
+
+    Our \texttt{output\_metric} of \rustdoc{metrics}{struct}{AbsoluteDistance} is compatible with any domain of the form \texttt{AllDomain(T)} where \texttt{T} has the trait \texttt{InfSub}, 
+    and our \texttt{output\_domain} of \texttt{AllDomain(TO)} is of this form and \texttt{TO} has the necessary trait.
+    Therefore our \texttt{input\_domain} and \texttt{input\_metric} are compatible.
+\end{proof}
+
+Before proceeding with proving the validity of the stability map, we provide a couple lemmas.
+
+\begin{lemma}
+    \label{dsym-sens}
+    $|\function(u) - \function(v)| \leq |\texttt{len(u)} - \texttt{len(v)}|$, 
+    where \texttt{len} is an alias for \\ \texttt{input\_domain.size}.
+\end{lemma}
+
+\begin{proof}
+    In line \ref{line:try-catch}, we know the argument to \texttt{TO.exact\_int\_cast} is non-negative and integral.
```

Comment:
The first sentence of this lemma was not immediately apparent the first time I read it. Maybe we should call out that `exact_int_cast` accepts an argument of type `usize` (or refer back to the pre-condition of the `ExactIntCast` trait) and therefore we know that the argument is non-negative and integral.

Reply:
I tied this reasoning to the interface of `CollectionDomain` instead!